In [ ]:
import time

In [ ]:
import random


def print_board(board):
    for row in board:
        print(" ".join(row))
    print()

def check_winner(board, player):
    for i in range(3):
        if all(board[i][j] == player for j in range(3)) or all(board[j][i] == player for j in range(3)):
            return True
    if all(board[i][i] == player for i in range(3)) or all(board[i][2 - i] == player for i in range(3)):
        return True
    return False

def is_board_full(board):
    return all(board[i][j] != ' ' for i in range(3) for j in range(3))

In [ ]:
def minimax(board, depth, is_maximizing):
    if check_winner(board, 'X'):
        return 1
    if check_winner(board, 'O'):
        return -1
    if is_board_full(board):
        return 0

    if is_maximizing:
        max_eval = float('-inf')
        for i in range(3):
            for j in range(3):
                if board[i][j] == ' ':
                    board[i][j] = 'X'
                    eval = minimax(board, depth + 1, False)
                    board[i][j] = ' '
                    max_eval = max(max_eval, eval)
        return max_eval
    else:
        min_eval = float('inf')
        for i in range(3):
            for j in range(3):
                if board[i][j] == ' ':
                    board[i][j] = 'O'
                    eval = minimax(board, depth + 1, True)
                    board[i][j] = ' '
                    min_eval = min(min_eval, eval)
        return min_eval

In [ ]:
def minimax(board, depth, is_maximizing, alpha, beta):
    if check_winner(board, 'X'):
        return 1
    if check_winner(board, 'O'):
        return -1
    if is_board_full(board):
        return 0

    if is_maximizing:
        max_eval = float('-inf')
        for i in range(3):
            for j in range(3):
                if board[i][j] == ' ':
                    board[i][j] = 'X'
                    eval = minimax(board, depth + 1, False, alpha, beta)
                    board[i][j] = ' '
                    max_eval = max(max_eval, eval)
                    alpha = max(alpha, eval)
                    if beta <= alpha:
                        break
        return max_eval
    else:
        min_eval = float('inf')
        for i in range(3):
            for j in range(3):
                if board[i][j] == ' ':
                    board[i][j] = 'O'
                    eval = minimax(board, depth + 1, True, alpha, beta)
                    board[i][j] = ' '
                    min_eval = min(min_eval, eval)
                    beta = min(beta, eval)
                    if beta <= alpha:
                        break
        return min_eval


In [ ]:
def best_move(board):
    best_eval = float('-inf')
    best_move = (-1, -1)
    alpha = float('-inf')
    beta = float('inf')

    for i in range(3):
        for j in range(3):
            if board[i][j] == ' ':
                board[i][j] = 'X'
                move_eval = minimax(board, 0, False, alpha, beta)
                #move_eval = minimax(board, 0, False)
                board[i][j] = ' '
                if move_eval > best_eval:
                    best_eval = move_eval
                    best_move = (i, j)

    return best_move


In [ ]:
def play_game():
    board = [[' ' for _ in range(3)] for _ in range(3)]
    current_player = 'O'

    while True:
        print_board(board)

        if current_player == 'X':
            print("Computer's Turn (X)")
            start_time = time.time()
            i, j = best_move(board)
            end_time = time.time()
            print("time taken is : ", end_time - start_time)
        else:
            print("Your Turn (O)")
            i, j = map(int, input("Enter row and column (0-2) separated by a space: ").split())

        if board[i][j] == ' ':
            board[i][j] = current_player

            if check_winner(board, current_player):
                print_board(board)
                if current_player == 'X':
                    print("Computer wins!")
                else:
                    print("You win!")
                break

            if is_board_full(board):
                print_board(board)
                print("It's a draw!")
                break

            current_player = 'X' if current_player == 'O' else 'O'
        else:
            print("Invalid move. Try again.")

In [ ]:
if __name__ == "__main__":
    play_game()

     
     
     

Your Turn (O)
Enter row and column (0-2) separated by a space: 0 0
O    
     
     

Computer's Turn (X)
time taken is :  0.7778761386871338
O    
  X  
     

Your Turn (O)
Enter row and column (0-2) separated by a space: 2 0
O    
  X  
O    

Computer's Turn (X)
time taken is :  0.03149294853210449
O    
X X  
O    

Your Turn (O)
Enter row and column (0-2) separated by a space: 1 2
O    
X X O
O    

Computer's Turn (X)
time taken is :  0.0013761520385742188
O X  
X X O
O    

Your Turn (O)
Enter row and column (0-2) separated by a space: 2 1
O X  
X X O
O O  

Computer's Turn (X)
time taken is :  0.00014209747314453125
O X  
X X O
O O X

Your Turn (O)
Enter row and column (0-2) separated by a space: 0 2
O X O
X X O
O O X

It's a draw!


In [ ]:
if __name__ == "__main__":
    play_game()

     
     
     

Your Turn (O)
Enter row and column (0-2) separated by a space: 0 0
O    
     
     

Computer's Turn (X)
time taken is :  0.17412757873535156
O    
  X  
     

Your Turn (O)
Enter row and column (0-2) separated by a space: 2 0
O    
  X  
O    

Computer's Turn (X)
time taken is :  0.01534891128540039
O    
X X  
O    

Your Turn (O)
Enter row and column (0-2) separated by a space: 1 2
O    
X X O
O    

Computer's Turn (X)
time taken is :  0.0012595653533935547
O X  
X X O
O    

Your Turn (O)
Enter row and column (0-2) separated by a space: 2 1
O X  
X X O
O O  

Computer's Turn (X)
time taken is :  9.250640869140625e-05
O X  
X X O
O O X

Your Turn (O)
Enter row and column (0-2) separated by a space: 0 2
O X O
X X O
O O X

It's a draw!


The Comparison of the timing for the different algorithm

(Timings are in Second)

| Step Taken | Min-Max  | Alpha-Beta Pruning |
|------------|----------|--------------------|
| (0, 0)     | 0.777    | 0.174              |
| (2, 0)     | 0.031    | 0.015              |
| (1, 2)     | 0.00137  | 0.00140            |
| (2, 1)     | 0.000142 | 0.00012            |

